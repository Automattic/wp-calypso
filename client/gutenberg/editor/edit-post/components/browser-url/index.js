/**
 * External dependencies
 */
import { Component } from '@wordpress/element';
import { withSelect } from '@wordpress/data';
import { addQueryArgs } from '@wordpress/url';

/**
 * Returns the Post's Edit URL.
 *
 * @param {number} postId Post ID.
 *
 * @return {string} Post edit URL.
 */
export function getPostEditURL( postId ) {
	return addQueryArgs( 'post.php', { post: postId, action: 'edit' } );
}

/**
 * Returns the Post's Trashedd URL.
 *
 * @param {number} postId    Post ID.
 * @param {string} postType Post Type.
 *
 * @return {string} Post trashed URL.
 */
export function getPostTrashedURL( postId, postType ) {
	return addQueryArgs( 'edit.php', {
		trashed: 1,
		post_type: postType,
		ids: postId,
	} );
}

export class BrowserURL extends Component {
	constructor() {
		super( ...arguments );

		this.state = {
			historyId: null,
		};
	}

	componentDidUpdate( prevProps ) {
		const { postId, postStatus, postType } = this.props;
		const { historyId } = this.state;

		if ( postStatus === 'trash' ) {
			this.setTrashURL( postId, postType );
			return;
		}

		if ( ( postId !== prevProps.postId || postId !== historyId ) && postStatus !== 'auto-draft' ) {
			this.setBrowserURL( postId );
		}
	}

	/**
	 * Navigates the browser to the post trashed URL to show a notice about the trashed post.
	 *
	 * @param {number} postId    Post ID.
	 * @param {string} postType  Post Type.
	 */
	setTrashURL( postId, postType ) {
		window.location.href = getPostTrashedURL( postId, postType );
	}

	/**
	 * Replaces the browser URL with a post editor link for the given post ID.
	 *
	 * Note it is important that, since this function may be called when the
	 * editor first loads, the result generated `getPostEditURL` matches that
	 * produced by the server. Otherwise, the URL will change unexpectedly.
	 *
	 * @param {number} postId Post ID for which to generate post editor URL.
	 */
	setBrowserURL( postId ) {
		window.history.replaceState(
			{ id: postId },
			'Post ' + postId,
			getPostEditURL( postId )
		);

		this.setState( () => ( {
			historyId: postId,
		} ) );
	}

	render() {
		return null;
	}
}

export default withSelect( ( select ) => {
	const { getCurrentPost } = select( 'core/editor' );
	const { id, status, type } = getCurrentPost();

	return {
		postId: id,
		postStatus: status,
		postType: type,
	};
} )( BrowserURL );
