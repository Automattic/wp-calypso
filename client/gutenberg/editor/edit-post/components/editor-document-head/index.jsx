/** @format */
/**
 * External dependencies
 */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { localize } from 'i18n-calypso';

/**
 * Internal dependencies
 */
import DocumentHead from 'components/data/document-head';
import { getPostType } from 'state/post-types/selectors';
import { getSelectedSiteId } from 'state/ui/selectors';

export const EditorDocumentHead = ( { isNewPost, postType, postTypeObject, translate } ) => {
	let title = '';
	if ( 'post' === postType ) {
		title = isNewPost ? translate( 'New Post' ) : translate( 'Edit Post' );
	} else if ( 'page' === postType ) {
		title = isNewPost ? translate( 'New Page' ) : translate( 'Edit Page' );
	} else if ( !! postTypeObject ) {
		title = isNewPost ? postTypeObject.labels.new_item : postTypeObject.labels.edit_item;
	} else {
		title = isNewPost ? translate( 'New' ) : translate( 'Edit' );
	}

	return <DocumentHead title={ title } />;
};

EditorDocumentHead.propTypes = {
	isNewPost: PropTypes.bool,
	postId: PropTypes.number,
	postType: PropTypes.string,
	postTypeObject: PropTypes.object,
};

const mapStateToProps = ( state, { postId, postType } ) => {
	const siteId = getSelectedSiteId( state );
	return {
		isNewPost: ! postId,
		postTypeObject: getPostType( state, siteId, postType ),
	};
};

export default connect( mapStateToProps )( localize( EditorDocumentHead ) );
