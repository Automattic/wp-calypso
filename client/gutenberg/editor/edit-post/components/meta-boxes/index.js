/**
 * External dependencies
 */
import React from 'react';

/**
 * WordPress dependencies
 */
import { withSelect } from '@wordpress/data';

/**
 * Internal dependencies
 */
import MetaBoxesArea from './meta-boxes-area';

function MetaBoxes( { location, isActive } ) {
	if ( ! isActive ) {
		return null;
	}

	return <MetaBoxesArea location={ location } />;
}

export default withSelect(
	( select, ownProps ) => ( {
		isActive: select( 'core/edit-post' ).getMetaBox( ownProps.location ).isActive,
	} ),
)( MetaBoxes );
