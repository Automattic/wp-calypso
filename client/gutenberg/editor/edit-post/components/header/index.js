/**
 * External dependencies
 */
import React from 'react';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { IconButton } from '@wordpress/components';
import {
	PostPreviewButton,
	PostSavedState,
	PostPublishPanelToggle,
} from '@wordpress/editor';
import { withDispatch, withSelect } from '@wordpress/data';
import { compose } from '@wordpress/compose';
import { DotTip } from '@wordpress/nux';

/**
 * Internal dependencies
 */
import MoreMenu from './more-menu';
import HeaderToolbar from './header-toolbar';
import PinnedPlugins from './pinned-plugins';

function Header( {
	isEditorSidebarOpened,
	openGeneralSidebar,
	closeGeneralSidebar,
	isPublishSidebarOpened,
	togglePublishSidebar,
} ) {
	const toggleGeneralSidebar = isEditorSidebarOpened ? closeGeneralSidebar : openGeneralSidebar;

	/* eslint-disable wpcalypso/jsx-classname-namespace */
	return (
		<div
			role="region"
			/* translators: accessibility text for the top bar landmark region. */
			aria-label={ __( 'Editor top bar' ) }
			className="edit-post-header"
			tabIndex="-1"
		>
			<HeaderToolbar />
			{ ! isPublishSidebarOpened && (
				<div className="edit-post-header__settings">
					<PostSavedState	/>
					<PostPreviewButton />
					<PostPublishPanelToggle
						isOpen={ isPublishSidebarOpened }
						onToggle={ togglePublishSidebar }
					/>
					<IconButton
						icon="admin-generic"
						label={ __( 'Settings' ) }
						onClick={ toggleGeneralSidebar }
						isToggled={ isEditorSidebarOpened }
						aria-expanded={ isEditorSidebarOpened }
					>
						<DotTip id="core/editor.settings">
							{ __( 'You’ll find more settings for your page and blocks in the sidebar. Click ‘Settings’ to open it.' ) }
						</DotTip>
					</IconButton>
					<PinnedPlugins.Slot />
					<MoreMenu />
				</div>
			) }
		</div>
	);
	/* eslint-enable wpcalypso/jsx-classname-namespace */
}

export default compose(
	withSelect( ( select ) => ( {
		isEditorSidebarOpened: select( 'core/edit-post' ).isEditorSidebarOpened(),
		isPublishSidebarOpened: select( 'core/edit-post' ).isPublishSidebarOpened(),
		hasBlockSelection: !! select( 'core/editor' ).getBlockSelectionStart(),
	} ) ),
	withDispatch( ( dispatch, { hasBlockSelection } ) => {
		const { openGeneralSidebar, closeGeneralSidebar, togglePublishSidebar } = dispatch( 'core/edit-post' );
		const sidebarToOpen = hasBlockSelection ? 'edit-post/block' : 'edit-post/document';
		return {
			openGeneralSidebar: () => openGeneralSidebar( sidebarToOpen ),
			closeGeneralSidebar: closeGeneralSidebar,
			togglePublishSidebar: togglePublishSidebar,
			hasBlockSelection: undefined,
		};
	} ),
)( Header );
