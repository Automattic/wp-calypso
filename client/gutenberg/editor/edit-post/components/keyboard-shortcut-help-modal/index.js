/** @format */
/* eslint-disable wpcalypso/jsx-classname-namespace */
/**
 * External dependencies
 */
import React from 'react';
import { castArray } from 'lodash';

/**
 * WordPress dependencies
 */
import { Fragment } from '@wordpress/element';
import { Modal, KeyboardShortcuts } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { rawShortcut } from '@wordpress/keycodes';
import { withSelect, withDispatch } from '@wordpress/data';
import { compose } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import shortcutConfig from './config';

const MODAL_NAME = 'edit-post/keyboard-shortcut-help';

const mapKeyCombination = keyCombination =>
	keyCombination.map( ( character, index ) => {
		if ( character === '+' ) {
			return <Fragment key={ index }>{ character }</Fragment>;
		}

		return (
			<kbd key={ index } className="edit-post-keyboard-shortcut-help__shortcut-key">
				{ character }
			</kbd>
		);
	} );

const ShortcutList = ( { shortcuts } ) => (
	<dl className="edit-post-keyboard-shortcut-help__shortcut-list">
		{ shortcuts.map( ( { keyCombination, description, ariaLabel }, index ) => (
			<div className="edit-post-keyboard-shortcut-help__shortcut" key={ index }>
				<dt className="edit-post-keyboard-shortcut-help__shortcut-term">
					<kbd
						className="edit-post-keyboard-shortcut-help__shortcut-key-combination"
						aria-label={ ariaLabel }
					>
						{ mapKeyCombination( castArray( keyCombination ) ) }
					</kbd>
				</dt>
				<dd className="edit-post-keyboard-shortcut-help__shortcut-description">{ description }</dd>
			</div>
		) ) }
	</dl>
);

const ShortcutSection = ( { title, shortcuts } ) => (
	<section className="edit-post-keyboard-shortcut-help__section">
		<h2 className="edit-post-keyboard-shortcut-help__section-title">{ title }</h2>
		<ShortcutList shortcuts={ shortcuts } />
	</section>
);

export function KeyboardShortcutHelpModal( { isModalActive, toggleModal } ) {
	const title = (
		<span className="edit-post-keyboard-shortcut-help__title">{ __( 'Keyboard Shortcuts' ) }</span>
	);

	return (
		<Fragment>
			<KeyboardShortcuts
				bindGlobal
				shortcuts={ {
					[ rawShortcut.access( 'h' ) ]: toggleModal,
				} }
			/>
			{ isModalActive && (
				<Modal
					className="edit-post-keyboard-shortcut-help"
					title={ title }
					closeLabel={ __( 'Close' ) }
					onRequestClose={ toggleModal }
				>
					{ shortcutConfig.map( ( config, index ) => (
						<ShortcutSection key={ index } { ...config } />
					) ) }
				</Modal>
			) }
		</Fragment>
	);
}

export default compose( [
	withSelect( select => ( {
		isModalActive: select( 'core/edit-post' ).isModalActive( MODAL_NAME ),
	} ) ),
	withDispatch( ( dispatch, { isModalActive } ) => {
		const { openModal, closeModal } = dispatch( 'core/edit-post' );

		return {
			toggleModal: () => ( isModalActive ? closeModal() : openModal( MODAL_NAME ) ),
		};
	} ),
] )( KeyboardShortcutHelpModal );
