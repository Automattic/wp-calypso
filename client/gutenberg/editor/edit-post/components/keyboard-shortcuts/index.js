/**
 * External dependencies
 */
import React from 'react';

/**
 * WordPress dependencies
 */
import { Component } from '@wordpress/element';
import { KeyboardShortcuts } from '@wordpress/components';
import { withSelect, withDispatch } from '@wordpress/data';
import { compose } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import shortcuts from '../../keyboard-shortcuts';

class EditorModeKeyboardShortcuts extends Component {
	constructor() {
		super( ...arguments );

		this.toggleMode = this.toggleMode.bind( this );
	}

	toggleMode() {
		const { mode, switchMode } = this.props;
		switchMode( mode === 'visual' ? 'text' : 'visual' );
	}

	render() {
		return (
			<KeyboardShortcuts
				bindGlobal
				shortcuts={ {
					[ shortcuts.toggleEditorMode.value ]: this.toggleMode,
				} }
			/>
		);
	}
}

export default compose( [
	withSelect( ( select ) => {
		return {
			mode: select( 'core/edit-post' ).getEditorMode(),
		};
	} ),
	withDispatch( ( dispatch ) => {
		return {
			switchMode: ( mode ) => {
				dispatch( 'core/edit-post' ).switchEditorMode( mode );
			},
		};
	} ),
] )( EditorModeKeyboardShortcuts );
