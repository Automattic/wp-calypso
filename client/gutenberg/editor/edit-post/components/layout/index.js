/** @format */
/**
 * External dependencies
 */
import React from 'react';
import classnames from 'classnames';

/**
 * WordPress dependencies
 */
import { Button, Popover, ScrollLock, navigateRegions } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import {
	AutosaveMonitor,
	UnsavedChangesWarning,
	EditorNotices,
	PostPublishPanel,
	PreserveScrollInReorder,
} from '@wordpress/editor';
import { withDispatch, withSelect } from '@wordpress/data';
import { Fragment } from '@wordpress/element';
import { PluginArea } from '@wordpress/plugins';
import { withViewportMatch } from '@wordpress/viewport';
import { compose } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import BrowserURL from 'gutenberg/editor/browser-url';
import Header from '../header';
import TextEditor from '../text-editor';
import VisualEditor from '../visual-editor';
import EditorModeKeyboardShortcuts from '../keyboard-shortcuts';
import SettingsSidebar from '../sidebar/settings-sidebar';
import Sidebar from '../sidebar';
import { PluginPostPublishPanel, PluginPrePublishPanel } from '@wordpress/edit-post';
import EditorRevisionsDialog from 'post-editor/editor-revisions/dialog';

function Layout( {
	mode,
	editorSidebarOpened,
	pluginSidebarOpened,
	publishSidebarOpened,
	hasFixedToolbar,
	closePublishSidebar,
	togglePublishSidebar,
	isMobileViewport,
	updatePost,
	post,
} ) {
	const sidebarIsOpened = editorSidebarOpened || pluginSidebarOpened || publishSidebarOpened;

	const className = classnames( 'edit-post-layout', {
		'is-sidebar-opened': sidebarIsOpened,
		'has-fixed-toolbar': hasFixedToolbar,
	} );

	const publishLandmarkProps = {
		role: 'region',
		/* translators: accessibility text for the publish landmark region. */
		'aria-label': __( 'Editor publish' ),
		tabIndex: -1,
	};

	const loadRevision = revision => {
		const { post_content: content, post_title: title, post_excerpt: excerpt } = revision;
		const postRevision = { ...post, content, title, excerpt };
		//TODO: what's the better action here? This only loads title, also tried other actions in core/editor without luck
		updatePost( postRevision );
	};

	/* eslint-disable wpcalypso/jsx-classname-namespace */
	return (
		<div className={ className }>
			<BrowserURL />
			<UnsavedChangesWarning />
			<AutosaveMonitor />
			<Header />
			<div
				className="edit-post-layout__content"
				role="region"
				/* translators: accessibility text for the content landmark region. */
				aria-label={ __( 'Editor content' ) }
				tabIndex="-1"
			>
				<EditorNotices />
				<PreserveScrollInReorder />
				<EditorModeKeyboardShortcuts />
				{ mode === 'text' && <TextEditor /> }
				{ mode === 'visual' && <VisualEditor /> }
			</div>
			<EditorRevisionsDialog loadRevision={ loadRevision } />
			{ publishSidebarOpened ? (
				<PostPublishPanel
					{ ...publishLandmarkProps }
					onClose={ closePublishSidebar }
					PrePublishExtension={ PluginPrePublishPanel.Slot }
					PostPublishExtension={ PluginPostPublishPanel.Slot }
				/>
			) : (
				<Fragment>
					<div className="edit-post-toggle-publish-panel" { ...publishLandmarkProps }>
						<Button
							isDefault
							type="button"
							className="edit-post-toggle-publish-panel__button"
							onClick={ togglePublishSidebar }
							aria-expanded={ false }
						>
							{ __( 'Open publish panel' ) }
						</Button>
					</div>
					<SettingsSidebar />
					<Sidebar.Slot />
					{ isMobileViewport && sidebarIsOpened && <ScrollLock /> }
				</Fragment>
			) }
			<Popover.Slot />
			<PluginArea />
		</div>
	);
	/* eslint-enable wpcalypso/jsx-classname-namespace */
}

export default compose(
	withSelect( select => ( {
		post: select( 'core/editor' ).getCurrentPost(),
		mode: select( 'core/edit-post' ).getEditorMode(),
		editorSidebarOpened: select( 'core/edit-post' ).isEditorSidebarOpened(),
		pluginSidebarOpened: select( 'core/edit-post' ).isPluginSidebarOpened(),
		publishSidebarOpened: select( 'core/edit-post' ).isPublishSidebarOpened(),
		hasFixedToolbar: select( 'core/edit-post' ).isFeatureActive( 'fixedToolbar' ),
	} ) ),
	withDispatch( dispatch => {
		const { updatePost } = dispatch( 'core/editor' );
		const { closePublishSidebar, togglePublishSidebar } = dispatch( 'core/edit-post' );
		return {
			closePublishSidebar,
			togglePublishSidebar,
			updatePost,
		};
	} ),
	navigateRegions,
	withViewportMatch( { isMobileViewport: '< small' } )
)( Layout );
