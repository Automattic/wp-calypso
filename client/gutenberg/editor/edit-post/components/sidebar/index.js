/**
 * WordPress Dependencies
 */
import { createSlotFill, withFocusReturn } from '@wordpress/components';
import { withSelect } from '@wordpress/data';
import { ifCondition, compose } from '@wordpress/compose';

/**
 * Internal Dependencies
 */

const { Fill, Slot } = createSlotFill( 'Sidebar' );

/**
 * Renders a sidebar with its content.
 *
 * @return {Object} The rendered sidebar.
 */
const Sidebar = ( { children, label } ) => {
	return (
		<Fill>
			<div
				className="edit-post-sidebar"
				role="region"
				aria-label={ label }
				tabIndex="-1"
			>
				{ children }
			</div>
		</Fill>
	);
};

const WrappedSidebar = compose(
	withSelect( ( select, { name } ) => ( {
		isActive: select( 'core/edit-post' ).getActiveGeneralSidebarName() === name,
	} ) ),
	ifCondition( ( { isActive } ) => isActive ),
	withFocusReturn,
)( Sidebar );

WrappedSidebar.Slot = Slot;

export default WrappedSidebar;
