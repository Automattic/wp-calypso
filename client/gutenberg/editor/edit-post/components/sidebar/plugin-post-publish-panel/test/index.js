/**
 * External dependencies
 */
import { mount } from 'enzyme';

/**
 * WordPress dependencies
 */
import { SlotFillProvider } from '@wordpress/components';

/**
 * Internal dependencies
 */
import PluginPostPublishPanel from '../';

jest.mock( '../../../../../packages/components/src/button' );

describe( 'PluginPostPublishPanel', () => {
	test( 'renders fill properly', () => {
		const wrapper = mount(
			<SlotFillProvider>
				<PluginPostPublishPanel
					className="my-plugin-post-publish-panel"
					title="My panel title"
					initialOpen={ true }
				>
					My panel content
				</PluginPostPublishPanel>
				<PluginPostPublishPanel.Slot />
			</SlotFillProvider>
		);

		expect( wrapper.find( 'Slot' ).children() ).toMatchSnapshot();
	} );
} );
