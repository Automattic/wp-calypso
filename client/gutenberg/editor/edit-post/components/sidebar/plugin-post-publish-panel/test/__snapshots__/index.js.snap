// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginPostPublishPanel renders fill properly 1`] = `
<div
  role="presentation"
>
  <PanelBody
    className="my-plugin-post-publish-panel"
    initialOpen={true}
    key="1---0/.0"
    title="My panel title"
  >
    <div
      className="components-panel__body my-plugin-post-publish-panel is-opened"
    >
      <h2
        className="components-panel__body-title"
      >
        <Button
          aria-expanded={true}
          className="components-panel__body-toggle"
          onClick={[Function]}
        >
          <button
            aria-expanded={true}
            className="components-button components-panel__body-toggle"
            onClick={[Function]}
            type="button"
          >
            <Dashicon
              className="components-panel__arrow"
              icon="arrow-up"
            >
              <svg
                aria-hidden={true}
                className="dashicon dashicons-arrow-up components-panel__arrow"
                focusable="false"
                height={20}
                role="img"
                viewBox="0 0 20 20"
                width={20}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 13l4.03-6L15 13H7z"
                />
              </svg>
            </Dashicon>
            My panel title
          </button>
        </Button>
      </h2>
      My panel content
    </div>
  </PanelBody>
</div>
`;
